core-bundle/src/Controller/AbstractController.php

```diff
@@ -72,4 +74,53 @@ protected function getCsrfFormOptions(): array
             'csrf_token_id' => $this->getParameter('contao.csrf_token_name'),
         ];
     }
+
+    /**
+     * Set the cache headers according to the page settings.
+     *
+     * @param Response $response The response object
+     *
+     * @return Response The response object
+     */
+    protected function setCacheHeaders(Response $response, PageModel $pageModel)
+    {
+        // Do not cache the response if caching was not configured at all or disabled explicitly
+        if ((false === $pageModel->cache || $pageModel->cache < 1) && (false === $pageModel->clientCache || $pageModel->clientCache < 1)) {
+            $response->headers->set('Cache-Control', 'no-cache, no-store');
+
+            return $response->setPrivate(); // Make sure the response is private
+        }
+
+        // Private cache
+        if ($pageModel->clientCache > 0) {
+            $response->setMaxAge($pageModel->clientCache);
+            $response->setPrivate(); // Make sure the response is private
+        }
+
+        // Shared cache
+        if ($pageModel->cache > 0) {
+            $response->setSharedMaxAge($pageModel->cache); // Automatically sets the response to public
+
+            // We vary on cookies if a response is cacheable by the shared
+            // cache, so a reverse proxy does not load a response from cache if
+            // the _request_ contains a cookie.
+            //
+            // This DOES NOT mean that we generate a cache entry for every
+            // response containing a cookie! Responses with cookies will always
+            // be private (@see Contao\CoreBundle\EventListener\MakeResponsePrivateListener).
+            //
+            // However, we want to be able to force the reverse proxy to load a
+            // response from cache, even if the request contains a cookie – in
+            // case the admin has configured to do so. A typical use case would
+            // be serving public pages from cache to logged in members.
+            if (!$pageModel->alwaysLoadFromCache) {
+                $response->setVary(['Cookie']);
+            }
+
+            // Tag the page (see #2137)
+            $this->tagResponse($pageModel);
```

Comment:
```suggestion
            $this->container->get('contao.cache.entity_tags')->tagWithModelInstance($pageModel);
```

Reply:
Changed in 0a0fdf4e4

Reply:
The change is wrong, I‘m afraid. Why did you not commit my suggestion?

Reply:
Fixed in 0181a84c8